compiler/optimizer/abstractinterpreter/InliningProposal.cpp

```diff
@@ -0,0 +1,260 @@
+
+      for (uint32_t i = 0; i < numChildren; i++)
+         idtNodeQueue.push_back(currentNode->getChild(i));
+
+      }
+
+   traceMsg(comp, "\n");
+   }
+
+
+void TR::InliningProposal::addNode(TR::IDTNode *node)
+   {
+   ensureBitVectorInitialized();
+
+   const int32_t index = node->getGlobalIndex() + 1;
+   if (_nodes->isSet(index))
+      {
+      return;
+      }
+
+   _nodes->set(index);
+
+   _cost = 0;
+   _benefit = 0;
+   }
+
+bool TR::InliningProposal::isEmpty()
+   {
+   if (!_nodes)
+      return true;
+
+   return _nodes->isEmpty();
+   }
+
+uint32_t TR::InliningProposal::getCost()
+   {
+   if (_cost == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _cost;
+   }
+
+double TR::InliningProposal::getBenefit()
+   {
+   if (_benefit == 0)
+      {
+      computeCostAndBenefit();
+      }
+
+   return _benefit;
+   }
+
+void TR::InliningProposal::computeCostAndBenefit()
+   {
+   _cost = 0;
+   _benefit = 0;
+
+   if (!_idt)
+      return;
+
+   TR_BitVectorIterator bvi(*_nodes);
+   int32_t idtNodeIndex;
+
+   while (bvi.hasMoreElements())
+      {
+      idtNodeIndex = bvi.getNextElement();
+      IDTNode *node = _idt->getNodeByGlobalIndex(idtNodeIndex - 1);
+      if (node == NULL)
+         {
+         continue;
+         }
+      _cost += node->getCost();
+      _benefit += node->getBenefit();
+      }
+   }
+
+void TR::InliningProposal::ensureBitVectorInitialized()
+   {
+   if (!_nodes)
+      _nodes = new (_region) TR_BitVector(_region);
+   }
+
+bool TR::InliningProposal::isNodeInProposal(TR::IDTNode* node)
+   {
+   if (node == NULL)
+      return false;
+   if (_nodes == NULL)
+      return false;
+   if (_nodes->isEmpty())
+      return false;
+
+   const int32_t idx = node->getGlobalIndex() + 1;
+
+   return _nodes->isSet(idx);
+   }
+
+void TR::InliningProposal::unionInPlace(TR::InliningProposal *a, TR::InliningProposal *b)
+   {
+   ensureBitVectorInitialized();
+   a->ensureBitVectorInitialized();
+   b->ensureBitVectorInitialized();
+
+   *_nodes = *a->_nodes;
+   *_nodes |= *b->_nodes;
+   _cost = 0;
+   _benefit = 0;
+   }
+
+bool TR::InliningProposal::intersects(TR::InliningProposal* other)
+   {
+   if (!_nodes || !other->_nodes)
+      return false;
+
+   return _nodes->intersects(*other->_nodes);
+   }
+
+
+TR::InliningProposalTable::InliningProposalTable(uint32_t rows, uint32_t cols, TR::Region& region) :
+      _rows(rows),
+      _cols(cols),
+      _region(region)
+   {
+   _table = new (region) InliningProposal**[rows];
+
+   for (uint32_t i = 0; i < rows; i ++)
+      {
+      _table[i] = new (region) InliningProposal*[cols];
+      memset(_table[i], 0, sizeof(InliningProposal*)*cols);
+      }
+   }
+
+TR::InliningProposal* TR::InliningProposalTable::get(uint32_t row, uint32_t col)
+   {
+   InliningProposal* proposal = NULL;
+
+   if (row <0 || col <0 || row >= _rows || col >= _cols)
+      proposal = getEmptyProposal();
```

Comment:
> I prefer to keep this implementation.

That was my suggestion too :slightly_smiling_face:

Thanks for the code comments!

> the out-of-bound indices are negative, and currently we let them overflow, which makes the algorithm more confusing. Would you suggest me to change all type of them to `int_32` or just use `static_cast<int>` to cast unsigned value to signed when calculating the index?

Initially I thought using signed integers wouldn't help, but after giving it some more thought, it seems to me that *all* out-of-bounds indices requested by the packing algorithm should be negative. So I think it makes sense to make all of the row/column indices/counts signed (`int32_t`), and assert that they're positive when creating the table. The `get()` method can assert the upper bounds and then return an empty proposal if either index is negative

By my reading we should be safe from signed overflow in the core algorithm. For row indices, it subtracts 1 at a time and I don't believe it goes beyond -1. For column indices the subtraction is `col - currentSet.getCost()`, where both operands are nonnegative, so that also doesn't overflow. There shouldn't be any overflow while computing `currentSet.getCost()` either - I believe that everywhere it occurs `currentSet` consists only of a single node (corresponding to `row`) and possibly some of its ancestors, so the total cost is bounded by the full budget. `newProposal->getCost()` should also be safe because the true total cost can exceed `col` only when `currentSet.getCost() > col`, in which case `newProposal` contains the same nodes as `currentSet`

Reply:
Yes, those out-of-bound indices are all negative, but actually, both row and column indices may overflow. For the row index, since the algorithm will check if the previous proposal is better, it will also do this when the row index is 0, which will cause the overflow. For the column index, `col` may be less than `currentSet.getCost()`, since this is a variety of DP algorithm, which will check each index in the table.

As a solution, I add a helper method, `TR::InliningProposal* getByOffset(uint32_t row, uint32_t rowOffset, uint32_t col, uint32_t colOffset);`. This method will first check if the offset are greater than the given index, if yes, they will return an empty proposal, elase it will call the original getter. I don't want to do explicit cast from `uint32_t` to `int32_t`, because this may cause overflow if the unsigned value exceeds `INT32_MAX`. This solution also avoid to change the index type to `int32_t`, which is wierd to have negative index.

Reply:
> `getByOffset` ... will first check if the offset are greater than the given index

Great idea :100: